pkg/tsdb/prometheus/prometheus.go

```diff
@@ -304,60 +286,49 @@ func (s *Service) parseQuery(queryContext *backend.QueryDataRequest, dsInfo *Dat
 		expr = strings.ReplaceAll(expr, "$__range", strconv.FormatInt(rangeS, 10)+"s")
 		expr = strings.ReplaceAll(expr, "$__rate_interval", intervalv2.FormatDuration(calculateRateInterval(interval, dsInfo.TimeInterval, s.intervalCalculator)))
 
-		if model.RangeQuery && model.InstantQuery {
-			return nil, fmt.Errorf("the provided query is not valid, expected only one of `range` and `instant` to be true")
-		}
-
-		var queryType PrometheusQueryType
-		var start time.Time
-		var end time.Time
-
-		if model.InstantQuery {
-			queryType = Instant
-			start = query.TimeRange.From
-			end = query.TimeRange.To
-		} else {
-			queryType = Range
-			// Align query range to step. It rounds start and end down to a multiple of step.
-			start = time.Unix(int64(math.Floor((float64(query.TimeRange.From.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
-			end = time.Unix(int64(math.Floor((float64(query.TimeRange.To.Unix()+model.UtcOffsetSec)/interval.Seconds()))*interval.Seconds()-float64(model.UtcOffsetSec)), 0)
+		rangeQuery := model.RangeQuery
+		if !model.InstantQuery && !model.RangeQuery {
+			// In older dashboards, we were not setting range query param and !range && !instant was run as range query
+			rangeQuery = true
 		}
 
 		qs = append(qs, &PrometheusQuery{
 			Expr:         expr,
 			Step:         interval,
 			LegendFormat: model.LegendFormat,
-			Start:        start,
-			End:          end,
+			Start:        query.TimeRange.From,
+			End:          query.TimeRange.To,
 			RefId:        query.RefID,
-			QueryType:    queryType,
+			InstantQuery: model.InstantQuery,
+			RangeQuery:   rangeQuery,
+			UtcOffsetSec: model.UtcOffsetSec,
 		})
 	}
-
 	return qs, nil
 }
 
-func parseResponse(value model.Value, query *PrometheusQuery) (data.Frames, error) {
+func parseResponse(value map[PrometheusQueryType]model.Value, query *PrometheusQuery) (data.Frames, error) {
 	frames := data.Frames{}
 
-	matrix, ok := value.(model.Matrix)
-	if ok {
-		matrixFrames := matrixToDataFrames(matrix, query)
-		frames = append(frames, matrixFrames...)
-	}
+	for queryType, value := range value {
+		matrix, ok := value.(model.Matrix)
+		if ok {
+			matrixFrames := matrixToDataFrames(matrix, query, queryType)
+			frames = append(frames, matrixFrames...)
```

Comment:
👍 sounds good, I'll update

Reply:
Actually after implementing adding of Meta info at 1 place, this is not possible because we need to add Meta info and append frames at the end of loop,  after we process response.

Reply:
Implemented here: https://github.com/grafana/grafana/pull/39462/commits/4b768fcc3d808fd63933cdb3b35942464faf5b73